libs/evcs-service/src/main/java/uk/gov/di/ipv/core/library/service/EvcsService.java

```diff
@@ -131,68 +133,67 @@ private void updateExistingUserVCs(
-                            .toList();
-            var userVCsToUpdate =
-                    Stream.concat(
-                                    Stream.concat(
-                                            existingCurrentUserVcsNotInSessionToUpdate.stream(),
-                                            existingPendingReturnUserVcsNotInSessionToUpdate
-                                                    .stream()),
-                                    existingPendingReturnUserVcsInSessionToUpdate.stream())
-                            .toList();
-
-            if (!userVCsToUpdate.isEmpty()) evcsClient.updateUserVCs(userId, userVCsToUpdate);
-        }
+        var existingPendingReturnUserVcsNotInSessionToUpdate =
+                existingEvcsUserVCs.stream()
+                        .filter(vc -> vc.state().equals(PENDING_RETURN))
+                        .filter(
+                                vc ->
+                                        credentials.stream()
+                                                .noneMatch(
+                                                        credential ->
+                                                                credential
+                                                                        .getVcString()
+                                                                        .equals(vc.vc())))
+                        .map(
+                                vc ->
+                                        new EvcsUpdateUserVCsDto(
+                                                getVcSignature(vc), EvcsVCState.ABANDONED, null));
+        Supplier<Stream<EvcsUpdateUserVCsDto>> existingPendingReturnUserVcsInSessionToUpdate =
+                () ->
+                        existingEvcsUserVCs.stream()
+                                .filter(vc -> vc.state().equals(PENDING_RETURN))
+                                .filter(
+                                        vc ->
+                                                credentials.stream()
+                                                        .anyMatch(
+                                                                credential ->
+                                                                        credential
+                                                                                .getVcString()
+                                                                                .equals(vc.vc())))
+                                .map(
+                                        vc ->
+                                                new EvcsUpdateUserVCsDto(
+                                                        getVcSignature(vc),
+                                                        EvcsVCState.CURRENT,
+                                                        null));
+        var existingCurrentUserVcsNotInSessionToUpdate =
+                existingPendingReturnUserVcsInSessionToUpdate.get().findAny().isPresent()
```

Comment:
I think this should be the `isPendingIdentity` parameter?

If it's a pending identity, then we only care about `existingPendingReturnUserVcsNotInSessionToUpdate`

If it's a current identity, then we _also_ care about `existingPendingReturnUserVcsInSessionToUpdate` and `existingCurrentUserVcsNotInSessionToUpdate`

You might find it easier to refactor like:

```
var vcUpdates = new ArrayList<EvcsUpdateUserVCsDto>();

// Pending VCs that are not in the session identity become abandoned
var existingPendingReturnUserVcsNotInSessionToUpdate = [...];
vcUpdates.addAll(existingPendingReturnUserVcsNotInSessionToUpdate)

if (!isPendingIdentity) {
  // Pending VCs that are in the session identity become current
  var existingPendingReturnUserVcsInSessionToUpdate = [...];
  vcUpdates.addAll(existingPendingReturnUserVcsInSessionToUpdate);

  // Current VCs that are not in the session identity become historic
  var existingCurrentUserVcsNotInSessionToUpdate = [...];
  vcUpdates.addAll(existingCurrentUserVcsNotInSessionToUpdate);
}

if (!vcUpdates.isEmpty()) evcsClient.updateUserVCs(userId, vcUpdates);
```